### ISSUE TITLE
Dropout support for SDPA with torch.compile

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#99905]
### Eager Mode
Currently when someone calls https://github.com/pytorch/pytorch/blob/f0e28b1cb9e0bc30877465b9c804a46461e537db/aten/src/ATen/native/native_functions.yaml#L14032 with `dropout_p != 0.0` in eager mode everything works great. FlashAttention is somewhat anomalous in its RNG handiling. The function will "checkpoint" the current state of the RNGEngine when calling the forward pass. It will NOT save the full attention mask for backward and instead save (for cuda) the philox_seed and philox_offset. When not using CudaGraphs these two int64_t can be used to fully capture the state of the PhiloxRNGEngine. The code that does this in flash_mha_fwd can be found https://github.com/pytorch/pytorch/blob/f0e28b1cb9e0bc30877465b9c804a46461e537db/aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp#L321

In the backward function a temporary PhiloxState is constructed from the seed and offset saved for bacwkard https://github.com/pytorch/pytorch/blob/f0e28b1cb9e0bc30877465b9c804a46461e537db/aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp#L498

Notice that I added a `TORCH_CHECK(at::cuda::currentStreamCaptureStatus() == at::cuda::CaptureStatus::None` This is because when using cuda_graphs more information that two ints is needed to fully rebuild the state.  Ideally it would be easy to pass around a convenient representation of the PhiloxState from forward to backward functions. I would like to find a solution for enabling FlashAttention with CudaGraphs.

### Torch Compile
There are two issues currently with using sdpa_flash_attention w/ `dropout_p != 0.0` and torch compile.
1.) The CudaGraph issue mentioned above.
2.) At the time of creating this PR: https://github.com/pytorch/pytorch/pull/92917 the two integers `philox offset` and `philox seed` were getting baked into the compiled graph.

For 2) I left a note in the meta_regsitrations.py file here: https://github.com/pytorch/pytorch/blob/39590d06c563d830d02b9f94611ab01f07133c97/torch/_meta_registrations.py#L2350

After the fact Ed and Alban mentioned that https://github.com/pytorch/pytorch/blob/39590d06c563d830d02b9f94611ab01f07133c97/torch/_meta_registrations.py#L2417 Could be replaced with a symint directly and that would likely fix the issue?

This code was added to enforce graph_break https://github.com/pytorch/pytorch/blob/39590d06c563d830d02b9f94611ab01f07133c97/torch/_dynamo/variables/torch.py#L513
Causing graph breaks is shown to have a noticable inmpact on the performance of a compiled module. I expereimented with NanoGPT and produced the following results:

Eager TImes
```
FlashAttention Kernel dropout=0.0: iter 3: loss 10.8671, time 902.01ms
FlashAttention Kernel dropout=0.1: iter 3: loss 10.8671, time 900.97ms
```
Compile Times
```
FlashAttention Kernel dropout=0.0 w/ torch compile: iter 3: loss 10.8674, time 552.06ms
FlashAttention Kernel dropout=0.1 w/ torch compile: iter 3: loss 10.8830, time 610.35ms
```

### Questions
I am planning on  upstreaming the recent changes from Xformers to mem-efficient attention. They have also enabled dropout support using a similar RNG checkpoint mechanism. I would like come up with a solution that better handles dropout with pt 2.0

How should I PhiloxState be passed from forward to backward functions? How can the meta function be updated to not bake in the offsets and seeds?  Can these changes be FC/BC?

cc @jbschlosser @bhosmer @cpuhrsch @erichan1 @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by eellison** [comment]
cc @anijain2305 - rng functionalization
